README.md
# face_recognition
The whole process of face authentication,Live detection, face registration and login,python ,tensorflow,opencv,dlib


changelog.md
[2019-12-13]
---
- 初步测试完成，train没跑完

[2019-12-12]
---
- 增加flask框架,能运行,暂未测试

test/test.py
#!/usr/bin/env python
#  -*- coding:utf-8 -*-

#import urllib2
#import ssl
import json
import requests
import time
#import httplib
#from socket import socket, AF_INET, SOCK_STREAM
#import urllib
#from urllib import urlencode

ipaddr="http://10.3.7.44:5000"

def req_post():
    url = ipaddr + "/test"

    headers = \
        {
            "Content-Type": "application/json",
        }
    body = \
        {
            "test1": "F:\\face_recognition\\face-recognization\\videos\\mouth_wuyuge.mp4",
            "test2":1383838438
        }
    r = requests.post(url, headers=headers, data=json.dumps(body))
    print (r.text)

def req_get():
    url = ipaddr + "/test?test1=F:\\face_recognition\\face-recognization\\videos\\mouth_wuyuge.mp4&test2=111"
    req = requests.get(url)
    print (req.text)

'''
def getone():
    url = ipaddr + "/getone?" \
          "url=/root/face-recognization/videos/eye_wuyuge.mp4" \
          "&" \
          "size=300"
    req = requests.get(url)
    print (req.text)

def getall():
    url = ipaddr + "/getall?" \
          "url=/root/face-recognization/videos/eye_wuyuge.mp4" \
          "&" \
          "username=wuyuge" \
          "&" \
          "size=300" \
          "&" \
          "number=100"
    req = requests.get(url)
    print (req.text)

def train():
    url = ipaddr + "/train"
    req = requests.get(url)
    print (req.text)

def findfaceold():
    url = ipaddr + "/findface?" \
          "url=/root/face-recognization/cache_face/1576204814577859.jpg"
    req = requests.get(url)
    print (req.text)

def liveeye():
    url = ipaddr + "/liveeye?" \
          "url=/root/face-recognization/videos/eye_wuyuge.mp4"
    req = requests.get(url)
    print (req.text)

def livemouth():
    url = ipaddr + "/livemouth?" \
          "url=/root/face-recognization/videos/mouth_wuyuge.mp4"
    req = requests.get(url)
    print (req.text)
'''

'''
def findface():
    url = ipaddr + "/find_face?" \
          "file=/root/face-recognization/cache_face/1576204814577859.jpg"\
          "&" \
          "tolerance=0.8" \

    req = requests.get(url)
    print (req.text)
'''

def findface():
    url = ipaddr + "/find_face?" \
          "file=./videos/mouth_wuyuge.mp4"
    req = requests.get(url)
    print (req.text)



def checkface():
    url = ipaddr + "/check_face?" \
          "file=./videos/mouth_wuyuge.mp4"
    req = requests.get(url)
    print (req.text)

def learnface():
    url = ipaddr + "/learn_face?" \
          "file=./videos/mouth_wuyuge.mp4"\
          "&" \
          "username=wuyuge" \
          
          
    req = requests.get(url,timeout=600)
    #req = requests.get(url,time.sleep(6))
    print (req.text)

if __name__ == '__main__':
    #req_post()
    #req_get()

    # getone()
    # getall()
    # train()
    # findfaceold()
    # liveeye()
    # livemouth()

    findface()
    #learnface()
    #learnface()

find_face.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import cv2
import dlib
import os
import random
import argparse
from scipy.spatial import distance as dist
from imutils import face_utils
import numpy as np
import imutils
import time
import argparse

import numpy as np
import tensorflow as tf




faces_my_path = 'cache_face'
size = 300
if not os.path.exists(faces_my_path):
    os.makedirs(faces_my_path)

ap = argparse.ArgumentParser()
ap.add_argument("-v", "--video", type=str, default="1.mp4",
                help="path to input video file")
args = vars(ap.parse_args())


def eye_aspect_ratio(eye):
   
    A = dist.euclidean(eye[1], eye[5])
    B = dist.euclidean(eye[2], eye[4])
    C = dist.euclidean(eye[0], eye[3])

    ear = (A + B) / (2.0 * C)

    return ear


def mouth_aspect_ratio(mouth):
    A = dist.euclidean(mouth[2], mouth[9])  # 51, 59
    B = dist.euclidean(mouth[4], mouth[7])  # 53, 57
    C = dist.euclidean(mouth[0], mouth[6])  # 49, 55

    mar = (A + B) / (2.0 * C)

    return mar



def live_detection_eye(url):
    status=1
    EYE_AR_THRESH = 0.2
    EYE_AR_CONSEC_FRAMES = 3
    COUNTER = 0
    TOTAL = 0
    detector = dlib.get_frontal_face_detector()
    predictor = dlib.shape_predictor("./models/shape_predictor_68_face_landmarks.dat")

    (lStart, lEnd) = face_utils.FACIAL_LANDMARKS_IDXS["left_eye"]
    (rStart, rEnd) = face_utils.FACIAL_LANDMARKS_IDXS["right_eye"]

    vs = cv2.VideoCapture(url)
    time.sleep(1.0)
    while True:
        success, frame = vs.read()
        if not success: break
        try:
            frame = imutils.resize(frame, width=450)
        except:
            continue
        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        rects = detector(gray, 0)
        for rect in rects:
            shape = predictor(gray, rect)
            shape = face_utils.shape_to_np(shape)
            leftEye = shape[lStart:lEnd]
            rightEye = shape[rStart:rEnd]
            leftEAR = eye_aspect_ratio(leftEye)
            rightEAR = eye_aspect_ratio(rightEye)

            ear = (leftEAR + rightEAR) / 2.0

            if ear < EYE_AR_THRESH:
                COUNTER += 1
            else:
                if COUNTER >= EYE_AR_CONSEC_FRAMES:
                    TOTAL += 1
                COUNTER = 0

        if TOTAL > 1:
            return status
            break
    if TOTAL < 1:
        status=status-1
        return status



def live_detection_mouth(url):
    status=1
    MAR_THRESH = 0.5
    MOUTH_AR_CONSEC_FRAMES = 3
    COUNTER = 0
    TOTAL = 0
    detector = dlib.get_frontal_face_detector()
    predictor = dlib.shape_predictor("./models/shape_predictor_68_face_landmarks.dat")
    (mStart, mEnd) = face_utils.FACIAL_LANDMARKS_IDXS["mouth"]
    vs = cv2.VideoCapture(url)
    time.sleep(2.0)
    while True:
        success, frame = vs.read()
        if not success: break
        try:
            frame = imutils.resize(frame, width=450)
        except:
            continue
        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        rects = detector(gray, 0)
        for rect in rects:
            shape = predictor(gray, rect)
            shape = face_utils.shape_to_np(shape)
            mouth = shape[mStart:mEnd]
            mar = mouth_aspect_ratio(mouth)
            #mouthHull = cv2.convexHull(mouth)
            #cv2.drawContours(frame, [mouthHull], -1, (0, 255, 0), 1)
            if mar < MAR_THRESH:
                COUNTER += 1
            else:
                if COUNTER >= MOUTH_AR_CONSEC_FRAMES:
                    TOTAL += 1
                COUNTER = 0

        if TOTAL > 1:
            return status
            break
    if TOTAL < 1:
        status=status-1
        return status
    
    
    
def load_graph(model_file):
    graph = tf.Graph()
    graph_def = tf.compat.v1.GraphDef()

    with open(model_file, "rb") as f:
        # graph_def.ParseFromString(f.read())
        graph_def.ParseFromString(f.read())
    with graph.as_default():
        tf.import_graph_def(graph_def)

    return graph


def read_tensor_from_image_file(file_name,
                                input_height=299,
                                input_width=299,
                                input_mean=0,
                                input_std=255):
    tf.compat.v1.disable_eager_execution()
    input_name = "file_reader"
    output_name = "normalized"
    file_reader = tf.io.read_file(file_name, input_name)
    if file_name.endswith(".png"):
        image_reader = tf.image.decode_png(
            file_reader, channels=3, name="png_reader")
    elif file_name.endswith(".gif"):
        image_reader = tf.squeeze(
            tf.image.decode_gif(file_reader, name="gif_reader"))
    elif file_name.endswith(".bmp"):
        image_reader = tf.image.decode_bmp(file_reader, name="bmp_reader")
    else:
        image_reader = tf.image.decode_jpeg(
            file_reader, channels=3, name="jpeg_reader")
    float_caster = tf.cast(image_reader, tf.float32)
    dims_expander = tf.expand_dims(float_caster, 0)
    resized = tf.compat.v1.image.resize_bilinear(dims_expander, [input_height, input_width])
    normalized = tf.divide(tf.subtract(resized, [input_mean]), [input_std])
    sess = tf.compat.v1.Session()
    result = sess.run(normalized)

    return result


def load_labels(label_file):
    label = []
    proto_as_ascii_lines = tf.io.gfile.GFile(label_file).readlines()
    for l in proto_as_ascii_lines:
        label.append(l.rstrip())
    return label

def get_faces(url):
    status=1
    file_name = url
    model_file = "./data_label/faces.pb"
    label_file = "./data_label/faces.txt"
    input_height = 299
    input_width = 299
    input_mean = 0
    input_std = 255
    input_layer = "Mul"
    output_layer = "final_result"

    graph = load_graph(model_file)
    t = read_tensor_from_image_file(
    file_name,
    input_height=input_height,
    input_width=input_width,
    input_mean=input_mean,
    input_std=input_std)

    input_name = "import/" + input_layer
    output_name = "import/" + output_layer
    input_operation = graph.get_operation_by_name(input_name)
    output_operation = graph.get_operation_by_name(output_name)

    with tf.compat.v1.Session(graph=graph) as sess:
        results = sess.run(output_operation.outputs[0], {
      input_operation.outputs[0]: t
    })
        
    
    results = np.squeeze(results)

    top_k = results.argsort()[-1:][::-1]
    user_name = load_labels(label_file)
    for i in top_k:
        if results[i] > 0.8:
            return status, user_name[i]
        else:
            status=status-1
            return status
            
    # print(labels[i],results[i])
    

check_face.py
# -*- coding: utf-8 -*-
"""
Created on Fri Apr 24 09:18:08 2020

@author: wangjingxian
"""

#python cap_one_image.py --video G:\python\face-recognization\videos\eye_wuyuge.mp4


import cv2
import dlib
import os
import random
import argparse
from scipy.spatial import distance as dist
from imutils.video import FileVideoStream
from imutils.video import VideoStream
from imutils import face_utils
import numpy as np
import argparse
import imutils
import time
from pathlib import Path

faces_my_path = '../cache/videos'
size = 300
if not os.path.exists(faces_my_path):
    os.makedirs(faces_my_path)

ap = argparse.ArgumentParser()
ap.add_argument("-v", "--video", type=str, default="1.mp4",
                help="path to input video file")
args = vars(ap.parse_args())


def cache_face(video):
    faces_my_path = '../cache/videos'
    size = 300
    if not os.path.exists(faces_my_path):
        os.makedirs(faces_my_path)

    detector = dlib.get_frontal_face_detector()
    cap = cv2.VideoCapture(args["video"])
    num = 1
    while True:
        if (num <= 1):
            success, img = cap.read()
            if success is True:
                gray_img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
            else:
                break
            dets = detector(gray_img, 1)
            for i, d in enumerate(dets):
                x1 = d.top() if d.top() > 0 else 0
                y1 = d.bottom() if d.bottom() > 0 else 0
                x2 = d.left() if d.left() > 0 else 0
                y2 = d.right() if d.right() > 0 else 0
                face = img[x1:y1, x2:y2]
                face = cv2.resize(face, (size, size))
                a=int(time.time() * 1000000)
                cv2.imwrite(faces_my_path + '/' + str(a) + '.jpg', face)
                print("======= "+faces_my_path + '/' + str(a) + '.jpg =======')
                num += 1

        else:

            break

def cap_one_image(url):
    size = 300
    faces_my_path = './cache/videos'
    if not os.path.exists(faces_my_path):
        os.makedirs(faces_my_path)

    detector = dlib.get_frontal_face_detector()
    cap = cv2.VideoCapture(url)
    num = 1
    status=1
    image = ""
    while True:
        if (num <= 1):
            success, img = cap.read()
            if success is True:
                gray_img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
                
            else:
                status=status-1
                break
            dets = detector(gray_img, 1)
            for i, d in enumerate(dets):
                x1 = d.top() if d.top() > 0 else 0
                y1 = d.bottom() if d.bottom() > 0 else 0
                x2 = d.left() if d.left() > 0 else 0
                y2 = d.right() if d.right() > 0 else 0
                face = img[x1:y1, x2:y2]
                face = cv2.resize(face, (size, size))
                a = int(time.time() * 1000000)
                cv2.imwrite(faces_my_path + '/' + str(a) + '.jpg', face)
                # print("======= " + faces_my_path + '/' + str(a) + '.jpg =======')
                #image = image + "======= " + faces_my_path + '/' + str(a) + '.jpg ======='
                image=faces_my_path + '/' + str(a) + '.jpg'
                num += 1

        else:
            break

    return status,image

if __name__ == "__main__":
    video = args["video"]
    cache_face(video)



main.py
# -*- coding: utf-8 -*-
"""
Created on Fri Apr 24 08:37:42 2020

@author: wangjingxian
"""

from flask import Flask, jsonify
from flask import request
from pathlib import Path
import json
import check_face as checkface
import learn_face as learnface
import find_face as findface
import socket

socket.setdefaulttimeout(600)


app = Flask(__name__)

@app.route('/check_face', methods=['GET'])
def check_face():
    if request.method == "GET":
        url = request.args.get("file")
        if checkurl(url) == False:
            js = {"status": 0}
            return json.dumps(js)
        status,image = checkface.cap_one_image(url)
        js = {"status": status,"image": image }
        return json.dumps(js)


@app.route('/learn_face', methods=['GET'])
def learn_face():
    if request.method == "GET":
        url = request.args.get("file")
        if checkurl(url) == False:
            js = {"status": 0}
            return json.dumps(js)
        username = request.args.get("username")
      
        status = learnface.dataset_construction(url,username)
        
        if status==1:
            #print('666666666666666666666666666')
            status = learnface.dataset_train()
            #print('7777777777777777777777777')
        js = {"status": status}
        return json.dumps(js)


@app.route('/find_face', methods=['GET'])
def find_face():
    if request.method == "GET":
        url = request.args.get("file")
        if checkurl(url) == False:
            js = {"status": 0}
            #print('aaaaaaaaaaaaaaaaaaaaaaaaa')
            return json.dumps(js)
        status1 = findface.live_detection_eye(url)
        #print('bbbbbbbbbbbbbbbbbbbb')
        status2 = findface.live_detection_mouth(url)
        
        username=""
        if status1==1 or status2==1:
            
            status,image=checkface.cap_one_image(url)
            
            if status==1:
                status,username = findface.get_faces(image)
            else:
                status=0
        else:
            status=0
        js = {"status": status, "user_name": username}
        return json.dumps(js)




def checkurl(url):
    my_file = Path(url)
    if my_file.exists():
        return True
    else:
        return False

if __name__ == '__main__':
    app.run(debug=True,port=5000,host="0.0.0.0")